### PR TITLE
[Skia] Fix corner cases involving drawing shadows

### PR DESCRIPTION
**Pull request by Scony**
#### 120f5b1d4f902ccd085929982aa6d9cf6bb8a4fb
<pre>
[Skia] Fix corner cases involving drawing shadows
<a href="https://bugs.webkit.org/show_bug.cgi?id=273239">https://bugs.webkit.org/show_bug.cgi?id=273239</a>

Reviewed by Adrian Perez de Castro.

So far - in the ports that use skia - shadows were being drawn in the same draw call as the original drawing operation.
This was possible as shadows were basically created using a filter that can be added to every drawing operation in skia.
The problem with such apporach is, however, that it does not work properly with some composite operations.
Some composite operations such as e.g. &apos;destination-atop&apos; are expected to alter drawing in a way that shadow
is being drawn in front of it&apos;s origin.
To achieve that with skia:
 - drawing of shadow must be done as a separate draw call.
 - drawing of both shadow and shadow origin must be done on separate transparency layers so that shadow, shadow origin,
   and the original content present before draw operation will overlap correctly.

* LayoutTests/fast/canvas/canvas-composite-fill-with-shadow-and-fitting-perfectly-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-composite-fill-with-shadow-and-fitting-perfectly.html: Added.
* LayoutTests/fast/canvas/canvas-composite-fill-with-shadow-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-composite-fill-with-shadow.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::beginCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::endCompositeLayer):
(WebCore::CanvasRenderingContext2DBase::fillRect):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.cpp:
(WebCore::BifurcatedGraphicsContext::beginTransparencyLayer):
* Source/WebCore/platform/graphics/BifurcatedGraphicsContext.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::beginTransparencyLayer):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::beginTransparencyLayer):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::BeginTransparencyLayerWithCompositeMode::apply const):
(WebCore::DisplayList::BeginTransparencyLayerWithCompositeMode::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::BeginTransparencyLayerWithCompositeMode::BeginTransparencyLayerWithCompositeMode):
(WebCore::DisplayList::BeginTransparencyLayerWithCompositeMode::compositeMode const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::beginTransparencyLayer):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordBeginTransparencyLayer):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.cpp:
(Nicosia::CairoOperationRecorder::beginTransparencyLayer):
* Source/WebCore/platform/graphics/nicosia/cairo/NicosiaCairoOperationRecorder.h:
* Source/WebCore/platform/graphics/skia/FontCascadeSkia.cpp:
(WebCore::FontCascade::drawGlyphs):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::drawRect):
(WebCore::GraphicsContextSkia::drawNativeImageInternal):
(WebCore::GraphicsContextSkia::drawSkiaPath):
(WebCore::GraphicsContextSkia::fillPath):
(WebCore::GraphicsContextSkia::strokePath):
(WebCore::GraphicsContextSkia::createDropShadowFilterIfNeeded const):
(WebCore::GraphicsContextSkia::drawOutsetShadow):
(WebCore::GraphicsContextSkia::drawSkiaRect):
(WebCore::GraphicsContextSkia::fillRect):
(WebCore::GraphicsContextSkia::beginTransparencyLayer):
(WebCore::GraphicsContextSkia::endTransparencyLayer):
(WebCore::GraphicsContextSkia::strokeRect):
(WebCore::GraphicsContextSkia::fillRoundedRectImpl):
(WebCore::GraphicsContextSkia::drawSkiaText):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
* Source/WebCore/rendering/GlyphDisplayListCache.cpp:
(WebCore::GlyphDisplayListCache::canShareDisplayList):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::beginTransparencyLayerWithCompositeMode):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordBeginTransparencyLayer):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Co-authored-by: Carlos Garcia Campos &lt;cgarcia@igalia.com&gt;
Canonical link: <a href="https://commits.webkit.org/280008@main">https://commits.webkit.org/280008@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/165007e17d2b0205dad75e7aa7f937209dbf91cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7961 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58383 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5831 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57399 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42127 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6122 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44612 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3986 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32723 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25740 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/54929 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29507 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5152 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3975 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51439 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5420 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59973 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30648 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5543 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52042 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31733 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47894 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51584 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12292 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32529 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31295 "Built successfully") | | |

### CODE REVIEW

**Comment by Scony** [comment]
To be continued once https://github.com/WebKit/WebKit/pull/27513 lands

**Comment by Scony** [comment]
The change in current form makes skia implementation to behave as close to CG/cairo as possible, although full alignment is not possible as CG and cairo work differently in some cases due to likely problems with spec, see https://github.com/whatwg/html/issues/10338

**Comment by csaavedra** [comment]
Can you rebase this? For some reason the base commit is not building in some bots.

File: Source/WebCore/platform/graphics/GraphicsContext.h [hunk 1]
```diff
@@ -324,6 +324,7 @@ class GraphicsContext {
     // Transparency Layers
 
     WEBCORE_EXPORT virtual void beginTransparencyLayer(float opacity);
+    virtual void beginTransparencyLayer(CompositeOperator, BlendMode = BlendMode::Normal);
```
Comment (carlosgcampos) [inline Source/WebCore/platform/graphics/GraphicsContext.h]: It seems this needs `WEBCORE_EXPORT` too

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.h [hunk 2]
```diff
@@ -686,6 +686,27 @@ class BeginTransparencyLayer {
...
+    CompositeOperator compositeOperator() const { return m_compositeOperator; }
+    BlendMode blendMode() const { return m_blendMode; }
+
+    WEBCORE_EXPORT void apply(GraphicsContext&) const;
+    void dump(TextStream&, OptionSet<AsTextFlag>) const;
+
+private:
+    CompositeOperator m_compositeOperator;
+    BlendMode m_blendMode;
+};
```
Comment (carlosgcampos) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.h]: Maybe we can make this shorter by using CompositeMode struct

File: Source/WebCore/platform/graphics/displaylists/DisplayListItems.h [hunk 3]
```diff
@@ -686,6 +686,24 @@ class BeginTransparencyLayer {
     float m_opacity;
 };
 
+class BeginTransparencyLayerWithCompositeMode {
+public:
+    static constexpr char name[] = "begin-transparency-layer-with-composite-mode";
+
+    BeginTransparencyLayerWithCompositeMode(CompositeMode compositeMode)
```
Comment (carlosgcampos) [inline Source/WebCore/platform/graphics/displaylists/DisplayListItems.h]: explicit

File: Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in [hunk 4]
```diff
@@ -55,6 +55,7 @@ messages -> RemoteDisplayListRecorder NotRefCounted Stream {
     DrawSystemImage(Ref<WebCore::SystemImage> systemImage, WebCore::FloatRect destinationRect)
     DrawPattern(WebCore::RenderingResourceIdentifier imageIdentifier, WebCore::FloatRect destRect, WebCore::FloatRect tileRect, WebCore::AffineTransform transform, WebCore::FloatPoint phase, WebCore::FloatSize spacing, struct WebCore::ImagePaintingOptions options)
     BeginTransparencyLayer(float opacity)
+    BeginTransparencyLayerWithCompositeMode(struct WebCore::CompositeMode compositeMode)
```
Comment (carlosgcampos) [inline Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:58]: This is called beginTransparencyLayer in RemoteDisplayListRecorder. So we should either rename it here, or if same name is not supported in IPC messages, rename the method name in the class.

File: Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp [hunk 5]
```diff
@@ -928,6 +998,43 @@ void GraphicsContextSkia::drawPattern(NativeImage& nativeImage, const FloatRect&
...
+    if (textDrawingMode().contains(TextDrawingMode::Stroke)) {
+        SkPaint paint = createStrokePaint();
+        setupStrokeSource(paint);
+        paint.setAntiAlias(enableAntialias);
+        m_canvas.drawTextBlob(blob, x, y, paint);
+    }
+
+    if (isVertical)
+        m_canvas.restore();
+
```
Comment (aperezdc) [inline Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp]: Let's remove this empty line before landing. Probably it ended up here when moving the existing code over from the other function.

**Review by aperezdc — APPROVED** [review]
Patch looks reasonable to me, I left only a very small nit.

FWIW, good idea reusing the cross-port entry points for setting the temporary composition layer; and thanks for taking the time to write the commit log notes&mdash;they do help to understand what's the intention 🙏🏼

**Pull request merged**